### DISCUSSION TITLE
Custom feature extractor

### DISCUSSION THREAD

**Discussion by jpcbertoldo** [open-edge-platform/anomalib#702] [Feature Requests]
I have a use case where I want to use a custom feature extractor (weights should be downloaded from a given url).
I'm looking for a generic feature extractor, the analogous of the folder dataset for feature extraction.

@ashwinvaidya17 I don't know what the torchfx is, but could your #675 eventually cover this use case?

If not, are you planning to merge it soon (so I can build on top of it)?

**Comment by ashwinvaidya17** [comment]
Sure! Do you want just the weights or even the ability to pass custom backbone architecture?

**Reply by jpcbertoldo** [reply]
I have a `model.pt` and a `module.py` where I can find the `PytorchModelClass` and I'd like to used it :)

# sketch

I'm thinking of something that would look like

```python

# get the model weights in a .pt
import torch
from pathlib import Path
fpathpt = Path("/path/to/model.pt")
pt = torch.load(fpathpt)

# load the module where the class is
import importlib.util
fpath_module = Path("/path/to/module.py")
spec = importlib.util.spec_from_file_location(fpath_module.stem, fpath_module)
module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(module)

# get the class and instantiate it
class_name = "PytorchModelClass"
model_class = getattr(module, class_name)
model_class_init_kwargs = dict(
    n_classes=16, 
    dev="cuda",
)
model = model_class(**model_class_init_kwargs)

# load the weights
model.load_state_dict(pt)
```

# api sketch

I suppose the API could look something like

```python
# ".pt" stands for PyTorch i think (?)
pt_feature_extractor = PytorchFeatureExtractor(
     weights_path="/path/to/model.pt",
     module_path="/path/to/module.py",
     class_name="PytorchModelClass",
     layers=[
            # not so sure how this part would work
           "path.to.submodule.in.the.model.0",
           "path.to.submodule.in.the.model.1",
     ],
)
```

# nice but not essential

1. downloading the weights from a given url
2. some sugar syntax for module+class+weight path like `/path/to/module.py:PytorchModelClass`

**Reply by jpcbertoldo** [reply]
@ashwinvaidya17 do you have an expected deadline for https://github.com/openvinotoolkit/anomalib/pull/675 ?

I'm just wondering because i'd really like to use the feature described above soon (not urgent but would be nice to have it soon).

I think it's probably better to leave this for another PR and build it on top of what you already did, no?

**Reply by ashwinvaidya17** [reply]
I've updated the PR. What do you think about the design? Here is an example of how it will work
```python
import CustomModel
from anomalib.models.components.feature_extractors import TorchFXFeatureExtractor
feature_extractor = TorchFXFeatureExtractor(
    CustomModel, ["linear_relu_stack.3"], weights="path/to/weights.pth"
)
input = torch.randn(1, 1, 28, 28)
features = feature_extractor(input)
print([layer for layer in features.keys()])
---
["linear_relu_stack.3"]
```

**Reply by jpcbertoldo** [reply]
Looks good but I in my use case I'd like to **(0)** instantiate my `CustomModel` from an arbitrary `.py` not necessarily in the path and doing so from the config file.

Does your design cover **(0)**? (I couldn't read the PR)

---

I guessed there would be some **(1)** dynamic loading of  module containing the class and **(2)** a dynamic instantiation of this (generic, arbitray) class.

Because of **(1)** I expected to have some args like `module_path="/path/to/module.py"` and `class_name="PytorchModelClass"`, and because of **(2)** I expected some arg like `init_args` (which would allow the user to properly instantiate from the config file). 

Does it make sense? 

---

If your design covers the case of using an instantiated `CustomModel` (i.e. steps **(1)** and **(2)** are assumed to be done) I think it's best to leave the rest (step **(0)**) to build later -- in any case I think `TorchFXFeatureExtractor` should **not** encapsulate the logic of the step **(0)**.

**Reply by ashwinvaidya17** [reply]
My design expects the class to be passed to the feature extractor and has support for extra kwargs to pass to the model when initializing it. But I'll discuss the design with the rest of the team to figure out what's the best approach here. If we deprecate TImm feature extractor then we can have more flexibility when designing torchfx feature extractor class. Otherwise we might have to ensure that both have similar parameters and scope.

**Reply by jpcbertoldo** [reply]
> My design expects the class to be passed to the feature extractor and has support for extra kwargs to pass to the model when initializing it

By initializing you mean `CustomClass(**init_args)`, right?
If yes, then we can figure out the dynamic load later : )

> If we deprecate TImm feature extractor then we can have more flexibility when designing torchfx feature extractor class. 

I don't see why deprecate timm, it's just a different use case (when the pretrained model is publicly available in timm).

> Otherwise we might have to ensure that both have similar parameters and scope.

Couldn't they have specific parameters?
I mean, sounds natural to me that such feature extractor (as I described before) would have more parameters.

**Reply by ashwinvaidya17** [reply]
> By initializing you mean CustomClass(**init_args), right?

Yes. In that case let's figure out dynamic loading later.

> I don't see why deprecate timm, it's just a different use case (when the pretrained model is publicly available in timm).

Maybe we can keep it.

> Couldn't they have specific parameters? I mean, sounds natural to me that such feature extractor (as I described before) would have more parameters.

Yes. Ideally it would be nice if we can just swap the feature extractors without changing how they are called. But I think it depends on how we decide to go forward. I have a feeling that we are going to end up with specific parameters between both as torchfx is more flexible.

**Reply by jpcbertoldo** [reply]
> Yes. Ideally it would be nice if we can just swap the feature extractors without changing how they are called. But I think it depends on how we decide to go forward. I have a feeling that we are going to end up with specific parameters between both as torchfx is more flexible.

Yes, sounds normal, the use cases just have different requirements :)

Tnx for clarifying ashwin!

**Comment by samet-akcay** [comment]
https://anomalib.readthedocs.io/en/latest/markdown/guides/how_to/models/feature_extractors.html
